aiohttp/payload.py

```diff
@@ -307,17 +307,39 @@ def __init__(
             if hdrs.CONTENT_DISPOSITION not in self.headers:
                 self.set_content_disposition(disposition, filename=self._filename)
 
+        self._writable = True
+        self._seekable = True
+        try:
+            # It is weird but some IO object dont have `seekable()` method as IOBase object,
+            # it seems better for us to direct try if the `seek()` and `tell()` is available
+            # e.g. tarfile.TarFile._Stream
```

Comment:
> This class doesn't appear to be an IOBase, so I'm wondering why it reaches this code. Does it match one of the payload subclasses?

You may take a look at `test_streaming_tarfile_on_redirect()` in `tests/test_formdata.py`, we have a `tf.extractfile(entry)` there, which will return a `tarfile.ExFileObject` and is inherited from `io.BufferedReader`. Its `__init__` calles `super().__init__(fileobj)`, which `fileobj` is with type `_FileInFile`, letting the `BufferedReader.raw` to be this `_FileInFile` object.

Note that the `seekable()` method in `_FileInFile` will called `self.fileobj.seekable()`, and as we are opening the tarfile with mode `r|`, the `fileobj` in `_FileInFile` will be `tarfile.TarFile._Stream` object. Therefore if we use `seekable()` here, it will raise `AttributeError` in the `test_streaming_tarfile_on_redirect()` test.

We can verify that with `print(tf.extractfile(entry).raw.fileobj)` at `test_streaming_tarfile_on_redirect()` in `tests/test_formdata.py`.

Reply:
> Note that the `seekable()` method in `_FileInFile` will called `self.fileobj.seekable()`, and as we are opening the tarfile with mode `r|`, the `fileobj` in `_FileInFile` will be `tarfile.TarFile._Stream` object. Therefore if we use `seekable()` here, it will raise `AttributeError` in the `test_streaming_tarfile_on_redirect()` test.

So, the file object we call does have a .seekable(), but it then tries to call it on another object which doesn't. That looks like an upstream bug. Should probably report that and get it fixed it..

Reply:
> > Note that the `seekable()` method in `_FileInFile` will called `self.fileobj.seekable()`, and as we are opening the tarfile with mode `r|`, the `fileobj` in `_FileInFile` will be `tarfile.TarFile._Stream` object. Therefore if we use `seekable()` here, it will raise `AttributeError` in the `test_streaming_tarfile_on_redirect()` test.
> 
> So, the file object we call does have a .seekable(), but it then tries to call it on another object which doesn't. That looks like an upstream bug. Should probably report that and get it fixed it..

I agree this is an upstream issue, but it is coming from the standard library `tarfile` (I found this issue in my local Python 3.10.12 and I haven't verified if this issue still exists in newer versions), which I think for compatibility we might need to keep this approach

Reply:
We can do a version check in order to remove the hack once cpython as fixed it.

Reply:
> We can do a version check in order to remove the hack once cpython as fixed it.

Just checked CPython GitHub for the 3.12 and 3.13 branches but seems the issue still exists. I think currently we may first add a note in this code section to inform the version check change is needed in the future

Reply:
Yep, but a fix may be backported, so would like to see if they'll get it merged quickly and then decide what we do here.

Reply:
Sure, but shouldn't expect to have an AttributeError, right? Should still look at getting that fixed upstream.

Reply:
> Sure, but shouldn't expect to have an AttributeError, right? Should still look at getting that fixed upstream.

After watching the upstream implementation, I believe it's hard to define `seekable` or not for `tarfile._Stream`, as it is only forward-seekable. I have opened an issue in the CPython repo (https://github.com/python/cpython/issues/124293), but I am not being positive for having a quick solution to it